ILLink fails on when only .NET Core 3 is installed (Microsoft.AspNetCore.Blazor.Mono)

Microsoft.AspNetCore.Blazor.Mono 0.8.0-preview1-20181126.1 contains a tool illink.dll which only runs on .NET Core 2.x. This means builds will fail on machines which only have .NET Core 3.x SDKs will fail.


```
2019-01-15T23:23:03.2879513Z   It was not possible to find any compatible framework version
2019-01-15T23:23:03.2881025Z   The specified framework 'Microsoft.NETCore.App', version '2.0.0' was not found.
2019-01-15T23:23:03.2886552Z     - Check application dependencies and target a framework version installed at:
2019-01-15T23:23:03.2892068Z         F:\vsagent\11\.dotnet\x64\
2019-01-15T23:23:03.2896169Z     - Installing .NET Core prerequisites might help resolve this problem:
2019-01-15T23:23:03.2896890Z         https://go.microsoft.com/fwlink/?LinkID=798306&clcid=0x409
2019-01-15T23:23:03.2897099Z     - The .NET Core framework and SDK can be installed from:
2019-01-15T23:23:03.2898295Z         https://aka.ms/dotnet-download
2019-01-15T23:23:03.2898690Z     - The following versions are installed:
2019-01-15T23:23:03.2899551Z         3.0.0-preview-27113-06 at [F:\vsagent\11\.dotnet\x64\shared\Microsoft.NETCore.App]
2019-01-15T23:23:03.2904079Z         3.0.0-preview-27219-3 at [F:\vsagent\11\.dotnet\x64\shared\Microsoft.NETCore.App]
2019-01-15T23:23:03.3300512Z F:\vsagent\11\s\src\Components\Blazor\Build\src\targets\Blazor.MonoRuntime.targets(439,5): error MSB3073: The command "dotnet "F:\vsagent\11\s\.nuget\packages\microsoft.aspnetcore.blazor.mono\0.8.0-preview1-20181126.1\build\netstandard1.0\../../tools/illink/illink.dll" -l none --verbose --strip-security true --exclude-feature com --exclude-feature sre -v false -c link -u link -b true -d "F:\vsagent\11\s\.nuget\packages\microsoft.aspnetcore.blazor.mono\0.8.0-preview1-20181126.1\build\netstandard1.0\../../tools/mono/bcl/" -d "F:\vsagent\11\s\.nuget\packages\microsoft.aspnetcore.blazor.mono\0.8.0-preview1-20181126.1\build\netstandard1.0\../../tools/mono/bcl/Facades/" -o "F:\vsagent\11\s\src\Components\Blazor\testassets\StandaloneApp\obj\Release\netstandard2.0\blazor\linker/" -x "F:\vsagent\11\s\src\Components\Blazor\Build\src\targets\BuiltInBclLinkerDescriptor.xml" -x "F:\vsagent\11\s\src\Components\Blazor\testassets\StandaloneApp\obj\Release\netstandard2.0\blazor\linker.descriptor.xml" -a "F:\vsagent\11\s\.nuget\packages\microsoft.extensions.dependencyinjection\3.0.0-preview.19059.4\lib\netstandard2.0\Microsoft.Extensions.DependencyInjection.dll" -a "F:\vsagent\11\s\.nuget\packages\microsoft.extensions.dependencyinjection.abstractions\3.0.0-preview.19059.4\lib\netstandard2.0\Microsoft.Extensions.DependencyInjection.Abstractions.dll" -a "F:\vsagent\11\s\.nuget\packages\microsoft.jsinterop\0.8.0-preview1-20181126.4\lib\netstandard2.0\Microsoft.JSInterop.dll" -a "F:\vsagent\11\s\.nuget\packages\mono.webassembly.interop\0.8.0-preview1-20181126.4\lib\netstandard2.0\Mono.WebAssembly.Interop.dll" -a "F:\vsagent\11\s\src\Components\Blazor\Blazor\src\bin\Release\netstandard2.0\Microsoft.AspNetCore.Blazor.dll" -a "F:\vsagent\11\s\src\Components\Browser\src\bin\Release\netstandard2.0\Microsoft.AspNetCore.Components.Browser.dll" -a "F:\vsagent\11\s\src\Components\Components\src\bin\Release\netstandard2.0\Microsoft.AspNetCore.Components.dll" -a "F:\vsagent\11\s\src\Components\Blazor\testassets\StandaloneApp\obj\Release\netstandard2.0\StandaloneApp.dll"" exited with code -2147450730. [F:\vsagent\11\s\src\Components\Blazor\testassets\StandaloneApp\StandaloneApp.csproj]

--- a/build/runtimes.props
+++ b/build/runtimes.props
@@ -1,6 +1,9 @@
 <Project>
 
   <ItemGroup>
+    <!-- BaselineGenerator is netcoreapp2.1. Other build tools may also require 2.x. -->
+    <DotNetCoreRuntime Include="2.2.1" />
+
     <DotNetCoreRuntime Include="$(MicrosoftNETCoreAppPackageVersion)"
       Feed="$(DotNetAssetRootUrl)"
       FeedCredential="$(DotNetAssetRootAccessTokenSuffix)" />

--- a/build/runtimes.props
+++ b/build/runtimes.props
@@ -1,9 +1,6 @@
 <Project>
 
   <ItemGroup>
-    <!-- Workaround https://github.com/aspnet/AspNetCore/issues/6726. Required because illink (part of Microsoft.AspNetCore.Blazor.Mono) depends on .NET Core 2.x -->
-    <DotNetCoreRuntime Include="2.2.1" />
-
     <DotNetCoreRuntime Include="$(MicrosoftNETCoreAppPackageVersion)"
       Feed="$(DotNetAssetRootUrl)"
       FeedCredential="$(DotNetAssetRootAccessTokenSuffix)" />

--- a/eng/Versions.props
+++ b/eng/Versions.props
@@ -164,7 +164,7 @@
     <SystemIdentityModelTokensJwtPackageVersion>5.3.0</SystemIdentityModelTokensJwtPackageVersion>
     <WindowsAzureStoragePackageVersion>8.1.4</WindowsAzureStoragePackageVersion>
     <!-- Dependencies for Blazor. -->
-    <MicrosoftAspNetCoreBlazorMonoPackageVersion>0.8.0-preview-20190204.1</MicrosoftAspNetCoreBlazorMonoPackageVersion>
+    <MicrosoftAspNetCoreBlazorMonoPackageVersion>0.10.0-preview-20190325.1</MicrosoftAspNetCoreBlazorMonoPackageVersion>
     <!-- Packages from 2.1/2.2 branches used for site extension build -->
     <MicrosoftAspNetCoreAzureAppServicesSiteExtension21PackageVersion>2.1.1</MicrosoftAspNetCoreAzureAppServicesSiteExtension21PackageVersion>
     <MicrosoftAspNetCoreAzureAppServicesSiteExtension22PackageVersion>2.2.0</MicrosoftAspNetCoreAzureAppServicesSiteExtension22PackageVersion>

--- a/src/Components/Blazor/Build/test/RuntimeDependenciesResolverTest.cs
+++ b/src/Components/Blazor/Build/test/RuntimeDependenciesResolverTest.cs
@@ -90,7 +90,7 @@ namespace Microsoft.AspNetCore.Blazor.Build.Test
                 "System.Data.dll",
                 "System.Diagnostics.Debug.dll",
                 "System.Diagnostics.Tracing.dll",
-                "System.Drawing.dll",
+                "System.Drawing.Common.dll",
                 "System.IO.Compression.dll",
                 "System.IO.Compression.FileSystem.dll",
                 "System.Linq.dll",

--- a/src/Components/Blazor/testassets/MonoSanityClient/Examples.cs
+++ b/src/Components/Blazor/testassets/MonoSanityClient/Examples.cs
@@ -11,6 +11,14 @@ namespace MonoSanityClient
 {
     public static class Examples
     {
+        static Examples()
+        {
+            // We have to populate GetHttpMessageHandler with something (like the real
+            // Blazor web assembly host does), otherwise HttpClientHandler's constructor
+            // gets into an infinite loop.
+            FakeHttpMessageHandler.Attach();
+        }
+
         public static string AddNumbers(int a, int b)
             => (a + b).ToString();
 

--- a/src/Components/Blazor/testassets/MonoSanityClient/FakeHttpMessageHandler.cs
+++ b/src/Components/Blazor/testassets/MonoSanityClient/FakeHttpMessageHandler.cs
@@ -1,0 +1,26 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Net.Http;
+using System.Reflection;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace MonoSanityClient
+{
+    class FakeHttpMessageHandler : HttpMessageHandler
+    {
+        public static void Attach()
+        {
+            var getHttpMessageHandlerField = typeof(HttpClient).GetField(
+                "GetHttpMessageHandler",
+                BindingFlags.Static | BindingFlags.NonPublic);
+            Func<HttpMessageHandler> handlerFactory = () => new FakeHttpMessageHandler();
+            getHttpMessageHandlerField.SetValue(null, handlerFactory);
+        }
+
+        protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+            => throw new NotImplementedException($"{nameof(FakeHttpMessageHandler)} cannot {nameof(SendAsync)}.");
+    }
+}

